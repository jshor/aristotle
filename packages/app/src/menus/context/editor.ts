import { Store } from 'pinia'
import DocumentState from '@/store/DocumentState'
import edit from '../submenus/edit'
import grid from '../submenus/grid'

export default function editor (store: Store<string, DocumentState, any>): MenuEntry[] {
  return edit(store, grid(store).concat({ type: 'separator' }))
}
