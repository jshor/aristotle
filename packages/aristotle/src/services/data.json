{
	"nodes": [{
		"id": "aqLvQGuk27",
		"type": "Switch",
		"name": "R",
		"x": 0,
		"y": 350
	}, {
		"id": "VEFvl5BKys",
		"type": "Switch",
		"name": "S",
		"x": 0,
		"y": 450
	}, {
		"id": "QzoKcbmstv",
		"type": "LogicGate",
		"name": "NOR_1",
		"subtype": "NOR",
		"x": 200,
		"y": 350
	}, {
		"id": "3OLPFz0Rzk",
		"type": "LogicGate",
		"name": "NOR_2",
		"subtype": "NOR",
		"x": 200,
		"y": 450
	}, {
		"id": "beG6WAra0l",
		"type": "LogicGate",
		"name": "NOR_3",
		"subtype": "NOR",
		"x": 400,
		"y": 350
	}, {
		"id": "c5XCJ9eill",
		"type": "Lightbulb",
		"name": "OUT_1",
		"x": 400,
		"y": 450
	}, {
		"id": "XNTLeHX6dZ",
		"type": "Lightbulb",
		"name": "OUT_2",
		"x": 400,
		"y": 500
	}],
	"connections": [{
		"inputId": "QzoKcbmstv",
		"outputId": "3OLPFz0Rzk",
		"sourceIndex": 0,
		"targetIndex": 1
	}, {
		"inputId": "VEFvl5BKys",
		"outputId": "3OLPFz0Rzk",
		"sourceIndex": 0,
		"targetIndex": 0
	}, {
		"inputId": "3OLPFz0Rzk",
		"outputId": "QzoKcbmstv",
		"sourceIndex": 0,
		"targetIndex": 1
	}, {
		"inputId": "aqLvQGuk27",
		"outputId": "QzoKcbmstv",
		"sourceIndex": 0,
		"targetIndex": 0
	}, {
		"inputId": "QzoKcbmstv",
		"outputId": "c5XCJ9eill",
		"sourceIndex": 0,
		"targetIndex": 0
	}, {
		"inputId": "3OLPFz0Rzk",
		"outputId": "XNTLeHX6dZ",
		"sourceIndex": 0,
		"targetIndex": 0
	}]
}
