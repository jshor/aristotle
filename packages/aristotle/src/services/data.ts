import uuid from '@/utils/uuid'

const R = uuid()
const S = uuid()
const NOR_1 = uuid()
const NOR_2 = uuid()
const NOR_3 = uuid()
const OUT_1 = uuid()
const OUT_2 = uuid()

const data = {
  nodes: [
    {
      id: R,
      type: 'Switch',
      name: 'R',
      x: 0,
      y: 350
    },
    {
      id: S,
      type: 'Switch',
      name: 'S',
      x: 0,
      y: 450
    },
    {
      id: NOR_1,
      type: 'LogicGate',
      name: 'NOR_1',
      x: 200,
      y: 350
    },
    {
      id: NOR_2,
      type: 'LogicGate',
      name: 'NOR_2',
      x: 200,
      y: 450
    },
    {
      id: NOR_3,
      type: 'LogicGate',
      name: 'NOR_3',
      x: 400,
      y: 350
    },
    {
      id: OUT_1,
      type: 'Lightbulb',
      name: 'OUT_1',
      x: 400,
      y: 450
    },
    {
      id: OUT_2,
      type: 'Lightbulb',
      name: 'OUT_2',
      x: 400,
      y: 500
    }
  ],
  connections: [
    {
      inputId: NOR_1,
      outputId: NOR_2,
      outputPortIndex: 1
    },
    {
      inputId: S,
      outputId: NOR_2,
      outputPortIndex: 0
    },
    {
      inputId: NOR_2,
      outputId: NOR_1,
      outputPortIndex: 1
    },
    {
      inputId: R,
      outputId: NOR_1,
      outputPortIndex: 0
    },
    {
      inputId: NOR_1,
      outputId: OUT_1,
      outputPortIndex: 0
    },
    {
      inputId: NOR_2,
      outputId: OUT_2,
      outputPortIndex: 0
    }
  ]
}

console.log(JSON.stringify(data))

export default data
