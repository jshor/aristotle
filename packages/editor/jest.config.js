module.exports = {
  collectCoverage: true,
  moduleFileExtensions: ['js'],
  transform: {
    '^.+\\.js$': 'babel-jest'
  },
  moduleNameMapper: {
    '^@/(.*)$': '<rootDir>/src/$1'
  },
  setupFiles: ['<rootDir>/test/setup.js']
}
