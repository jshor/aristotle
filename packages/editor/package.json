{
  "name": "@aristotle/editor",
  "version": "1.0.0",
  "main": "dist/editor.js",
  "scripts": {
    "dev": "webpack -w",
    "build": "webpack",
    "test": "vue-cli-service test:unit",
    "lint": "vue-cli-service lint"
  },
  "license": "MIT",
  "dependencies": {
    "@aristotle/logic-circuit": "1.0.0",
    "@aristotle/logic-gates": "1.0.0",
    "draw2d": "^1.0.19"
  },
  "devDependencies": {
    "@babel/core": "^7.4.3",
    "@babel/plugin-proposal-class-properties": "^7.4.4",
    "@babel/preset-env": "^7.4.3",
    "babel-jest": "^24.7.1",
    "babel-loader": "^8.0.6",
    "jest": "^24.7.1",
    "jquery": "^3.3.1",
    "jquery-ui": "^1.12.1",
    "raw-loader": "^2.0.0",
    "webpack": "^4.31.0",
    "webpack-cli": "^3.3.2"
  }
}
