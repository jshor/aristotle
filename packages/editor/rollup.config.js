import commonjs from 'rollup-plugin-commonjs'

export default {
  input: 'src/index.js',
  output: {
    file: 'dist/index.js',
    format: 'umd',
    name: 'editor',
    exports: 'named'
  },
  plugins: [ commonjs() ]
}