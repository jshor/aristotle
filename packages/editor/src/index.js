import 'jquery-ui/ui/widgets/draggable'
import 'jquery-ui/ui/widgets/droppable'
import 'jquery-ui-touch-punch'

export { default as Editor } from './Editor'
export { default as CommandModel } from './models/CommandModel'
export { default as EditorModel } from './models/EditorModel'
