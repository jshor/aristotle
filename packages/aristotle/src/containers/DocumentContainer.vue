<template>
  <document>
    <template v-slot:editor>
      <div
        :id="document.id"
        :style="{
          width: '4998px',
          height: '4998px'
        }"
      />
    </template>
    
    <template v-slot:oscilloscope>
      <div class="oscilloscope-inner">OSCILLOSCOPE YAY</div>
    </template>
  </document>
</template>

<script>
import { mapGetters, mapState } from 'vuex'
import Toolbar from '@/components/Toolbar'
import Toolbox from '@/components/Toolbox'
import Document from '@/components/Document'
import { Editor, CommandModel } from '@aristotle/editor'
import DocumentModel from '@/models/DocumentModel'

export default {
  name: 'DocumentContainer',
  components: {
    Editor,
    Toolbar,
    Toolbox,
    Document
  },
  data () {
    return {
      canvas: null
    }
  },
  props: {
    document: {
      type: DocumentModel,
      required: true
    }
  },
  computed: {
    ...mapState(['relayedCommand'])
  },
  watch: {
    relayedCommand: {
      deep: true,
      handler (command) {
        this.canvas.applyCommand(command)
      }
    }
  },
  methods: {
    onRelayCommand ({ command, payload }) {
      this.$store.commit('RELAY_COMMAND', new CommandModel(command, payload))
    },
    onUpdateEditor (editorModel) {
      this.$store.commit('SET_EDITOR_MODEL', editorModel)
    },
    toolboxChanged (payload) {
      this.onRelayCommand({ command: 'UPDATE_ELEMENT', payload })
    },
    closeToolbox () {
      this.$store.commit('SET_TOOLBOX_VISIBILITY', false)
    },
    pan () {
      this.canvas.setMouseMode('PANNING')
    },
    select () {
      this.canvas.setMouseMode('SELECTION')
    },
    step () {
      this.canvas.step()
    },
    onCanvasUpdate (canvas) {
      const model = canvas.getEditorModel() // should be v-model (emit `value`)

      this.$store.commit('SET_EDITOR_MODEL', model)
      this.$store.commit('SET_TOOLBOX_VISIBILITY', false)
    },
    onToolbox (editor, settings) {
      this.$store.commit('SET_TOOLBOX_SETTINGS', settings)
      this.$store.commit('SET_TOOLBOX_VISIBILITY', true)
    },
    hideToolbox () {
    }
  },
  mounted () {
    setTimeout(() => {
      // TODO: Editor needs to be created AFTER the toolboxContainer is ready
      // because it looks for all .ui-droppable elements to apply d-n-d ops to.
      // maybe create an action for this that is called in mounted() of ToolboxContainer?
      this.canvas = new Editor(this.document.id.toString())

      this.canvas.on('toolbox', this.onToolbox)
      this.canvas.on('select', () => this.onCanvasUpdate(this.canvas))
      this.canvas.on('deselect', () => this.onCanvasUpdate(this.canvas))
      this.canvas.on('commandStackChanged', () => this.onCanvasUpdate(this.canvas))
      this.canvas.on('zoomed', (editor, { value }) => {
        this.$store.commit('SET_ZOOM_FACTOR', value)
      })
      this.canvas.load(this.document.data)
    })
  }
}
</script>

<style>
.oscilloscope-inner {
  background-color: #1C1D24;
  height: 100%;
  box-sizing: border-box;
}
</style>