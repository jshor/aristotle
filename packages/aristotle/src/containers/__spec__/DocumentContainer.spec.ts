import { createLocalVue, shallowMount } from '@vue/test-utils'
import { ICommand } from '@aristotle/editor'
import Vuex from 'vuex'
import DocumentContainer from '../DocumentContainer.vue'
import data from '../../mocks/document.json'

const localVue = createLocalVue()

localVue.use(Vuex)

jest.mock('@aristotle/editor', () => {
  class Editor {
    load = jest.fn()
    on = jest.fn()
    getEditorModel = jest.fn()
    applyCommand = jest.fn()
  }

  return { Editor }
})

describe('Document Container', () => {
  const documentId = 'test123'
  let wrapper, state

  const createWrapper = () => {
    state = {
      activeDocumentId: documentId
    }

    return shallowMount(DocumentContainer, {
      propsData: {
        document: {
          id: documentId,
          data,
          editorModel: {}
        }
      },
      localVue,
      store: new Vuex.Store({
        state,
        actions: {
          relayCommand: jest.fn()
        }
      })
    })
  }

  beforeEach(() => {
    wrapper = createWrapper()
  })

  // it('should match the snapshot', () => {
  //   expect(wrapper.html()).toMatchSnapshot()
  // })

  describe('applyCommand()', () => {
    const command: ICommand = {
      type: 'TestCommand',
      payload: null
    }

    // it('should call `applyCommand()` when the `relayedCommand` property changes', () => {
    //   jest.spyOn(wrapper.vm.editor, 'applyCommand')

    //   state.relayedCommand = null
    //   state.relayedCommand = {
    //     ...command,
    //     documentId
    //   }

    //   expect.assertions(1)
    //   Vue.nextTick(() => {
    //     expect(wrapper.vm.editor.applyCommand).toHaveBeenCalledTimes(1)
    //   })
    // })

    it('should send the command to the editor if the document is active', () => {
      jest.spyOn(wrapper.vm.editor, 'applyCommand')

      wrapper.vm.applyCommand(command)

      expect(wrapper.vm.editor.applyCommand).toHaveBeenCalledTimes(1)
      expect(wrapper.vm.editor.applyCommand).toHaveBeenCalledWith(command)
    })

    it('should not send the command to the editor if the document is inactive', () => {
      state.activeDocumentId = 'unknown-id'

      jest.spyOn(wrapper.vm.editor, 'applyCommand')

      wrapper.vm.applyCommand(command)

      expect(wrapper.vm.editor.applyCommand).not.toHaveBeenCalled()
    })
  })

  describe('setActivity()', () => {
    describe('when the document is active', () => {
      beforeEach(() => {
        jest.spyOn(wrapper.vm, 'relayCommand')
      })

      it('should pass the SET_ACTIVITY command with true if the document is focused', () => {
        jest
          .spyOn(document, 'hasFocus')
          .mockReturnValue(true)

        wrapper.vm.setActivity()

        expect(wrapper.vm.relayCommand).toHaveBeenCalledTimes(1)
      })

      it('should pass the SET_ACTIVITY command with true if the document is focused', () => {
        jest
          .spyOn(document, 'hasFocus')
          .mockReturnValue(false)

        wrapper.vm.setActivity()

        expect(wrapper.vm.relayCommand).toHaveBeenCalledTimes(1)
      })
    })

    describe('when the document is inactive', () => {
      it('should pass the SetActivity command with false', () => {
        state.activeDocumentId = 'unknown-id'

        jest
          .spyOn(document, 'hasFocus')
          .mockReturnValue(true)
        jest
          .spyOn(wrapper.vm, 'relayCommand')

        wrapper.vm.setActivity()

        expect(wrapper.vm.relayCommand).toHaveBeenCalledTimes(1)
        expect(wrapper.vm.relayCommand).toHaveBeenCalledWith({
          type: 'SetActivity',
          payload: false,
          documentId: wrapper.vm.document.id
        })
      })
    })
  })
})
