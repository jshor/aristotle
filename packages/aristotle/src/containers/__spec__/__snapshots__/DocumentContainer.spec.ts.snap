// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document Container should match the snapshot 1`] = `
<document-stub oscilloscopeenabled="true">
  <div style="width: 4998px; height: 4998px;"></div>
  <!---->
  <div class="zoom"><button class="zoom__out"><i class="fas fa-search-minus"></i></button>
    <div class="zoom__level">1</div> <button class="zoom__out"><i class="fas fa-search-plus"></i></button>
  </div>
  <oscilloscope-container-stub waves="[object Object]"></oscilloscope-container-stub>
</document-stub>
`;
